File: SWEA/D4/1218. ［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기/［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기.cpp
#include<iostream>
#include<stack>

using namespace std;
stack<char> stk;
bool pairCheck(char toStack, char inStack) {
	if (toStack == ')' && inStack == '(') return true;
	else if (toStack == ']' && inStack == '[') return true;
	else if (toStack == '}' && inStack == '{') return true;
	else if (toStack == '>' && inStack == '<') return true;
	return false;
}
int main(int argc, char** argv)
{
	int len, test_case;
	for (test_case = 1; test_case < 11; test_case++) {
		string str;
		cin >> len >> str;
		int ans = 1;
		for (char c : str) {
			if (c == '(' || c == '{' || c == '[' || c == '<')
				stk.push(c);
			else {
				if (!pairCheck(c, stk.top())) {
					ans = 0;
					break;
				}
				stk.pop();
			}
		}
		cout << '#' << test_case << ' ' << ans << '\n';
		while (!stk.empty()) {
			stk.pop();
		}
	}
	return 0;//정상종료시 반드시 0을 리턴해야합니다.
}

File: SWEA/D4/1218. ［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기/［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기.java
import java.util.*;
import java.io.*;

class Solution
{
    static Stack<Character> stk;
	public static void main(String args[]) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        for(int test=1;test<11;test++) {
            int len = Integer.parseInt(br.readLine());
            stk = new Stack<Character>();
            int ans = 1;

            String str = br.readLine();

            for (int i = 0; i < len; i++) {
                char ch = str.charAt(i);
                if(ch == '(' || ch == '{' || ch == '[' || ch == '<')
                    stk.push(ch);
                else {
                    if(!pairCheck(ch, stk.pop())) {
                        ans = 0;
                        break;
                    }
                }
            }
            System.out.println("#" + test + " " + ans);
        }
        br.close();
    }
    public static boolean pairCheck(char toStack, char inStack) {
        if(toStack == ')' && inStack == '(') return true;
        else if(toStack == '}' && inStack == '{') return true;
        else if(toStack == ']' && inStack == '[') return true;
        else if(toStack == '>' && inStack == '<') return true;
        else return false;
    }
}